File: desktop/src/types/cron.ts
/**
 * Cron job types matching cron/jobs.py.
 * @source cron/jobs.py
 */

/** Schedule kind. */
export type ScheduleKind = 'once' | 'interval' | 'cron';

/** Schedule for a cron job. */
export interface Schedule {
  kind: ScheduleKind;
  expr?: string;
  minutes?: number;
  run_at?: string;
  display: string;
}

/** Repeat configuration for a cron job. */
export interface Repeat {
  times: number | null;
  completed: number;
}

/** Delivery kind. */
export type DeliveryKind = 'origin' | 'local' | string;

/** Delivery configuration for a cron job. */
export interface Delivery {
  kind: DeliveryKind;
  channels?: string[];
}

/** Cron job as stored in jobs.json. */
export interface CronJob {
  id: string;
  name: string;
  prompt: string;
  skills: string[];
  skill: string | null;
  model: string | null;
  provider: string | null;
  base_url: string | null;
  api_key?: string | null;
  script: string | null;
  schedule: Schedule;
  schedule_display: string;
  repeat: Repeat;
  enabled: boolean;
  state: 'scheduled' | 'paused' | 'completed' | 'running';
  paused_at: string | null;
  paused_reason: string | null;
  created_at: string;
  next_run_at: string | null;
  last_run_at: string | null;
  last_status: 'ok' | 'error' | null;
  last_error: string | null;
  last_delivery_error: string | null;
  deliver: DeliveryKind;
  origin: Record<string, unknown> | null;
}

/** Cron job creation parameters. */
export interface CreateCronJobParams {
  prompt: string;
  schedule: string;
  name?: string;
  repeat?: number | null;
  deliver?: DeliveryKind | null;
  origin?: Record<string, unknown> | null;
  skill?: string | null;
  skills?: string[] | null;
  model?: string | null;
  provider?: string | null;
  base_url?: string | null;
  script?: string | null;
}

/** Cron job update parameters. */
export interface UpdateCronJobParams {
  name?: string;
  prompt?: string;
  schedule?: string;
  enabled?: boolean;
  repeat?: number | null;
  deliver?: DeliveryKind | null;
  skill?: string | null;
  skills?: string[] | null;
  model?: string | null;
  provider?: string | null;
  base_url?: string | null;
  script?: string | null;
}

File: desktop/sidecar/daemon/schemas/cron.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


class CronOverlay(BaseModel):
    pinned: bool = False
    color: Optional[str] = None
    note: Optional[str] = None
    updated_at: Optional[str] = None


class MergedCronJob(BaseModel):
    id: str
    schedule: str
    prompt: str
    enabled: bool
    created_at: str
    name: Optional[str] = None
    skills: list[str] = Field(default_factory=list)
    skill: Optional[str] = None
    model: Optional[str] = None
    provider: Optional[str] = None
    base_url: Optional[str] = None
    script: Optional[str] = None
    schedule_display: Optional[str] = None
    repeat: Optional[dict[str, Any]] = None
    state: Optional[str] = None
    paused_at: Optional[str] = None
    paused_reason: Optional[str] = None
    next_run_at: Optional[str] = None
    last_run_at: Optional[str] = None
    last_status: Optional[str] = None
    last_error: Optional[str] = None
    last_delivery_error: Optional[str] = None
    deliver: Optional[str] = None
    origin: Optional[dict[str, Any]] = None
    desktop: CronOverlay = Field(default_factory=CronOverlay)


class CreateCronJobRequest(BaseModel):
    prompt: str = Field(min_length=1)
    schedule: str = Field(min_length=1)
    name: Optional[str] = None
    repeat: Optional[int] = None
    deliver: Optional[str] = None
    origin: Optional[dict[str, Any]] = None
    skill: Optional[str] = None
    skills: Optional[list[str]] = None
    model: Optional[str] = None
    provider: Optional[str] = None
    base_url: Optional[str] = None
    script: Optional[str] = None


class UpdateCronJobRequest(BaseModel):
    name: Optional[str] = None
    prompt: Optional[str] = None
    schedule: Optional[str] = None
    enabled: Optional[bool] = None
    repeat: Optional[int] = None
    deliver: Optional[str] = None
    skill: Optional[str] = None
    skills: Optional[list[str]] = None
    model: Optional[str] = None
    provider: Optional[str] = None
    base_url: Optional[str] = None
    script: Optional[str] = None
    paused_at: Optional[str] = None
    paused_reason: Optional[str] = None
